c2rust-analyze/src/main.rs

```diff
@@ -474,6 +503,29 @@ fn run(tcx: TyCtxt) {
                 || info.contains(PointerInfo::NOT_TEMPORARY_REF))
     }
 
+    // track all types mentioned in extern blocks, we
+    // don't want to rewrite those
+    let mut foreign_mentioned_tys = HashSet::new();
+    for item in tcx.hir_crate_items(()).foreign_items() {
+        let did = item.def_id.to_def_id();
+        match tcx.def_kind(did) {
+            DefKind::Fn | DefKind::AssocFn => {
+                let sig = tcx.fn_sig(did);
+                let sig = tcx.erase_late_bound_regions(sig);
+                for ty in sig.inputs_and_output {
+                    walk_args_and_fields(tcx, ty, &mut |did| foreign_mentioned_tys.insert(did));
+                }
```

Comment:
Once the above comments are addressed an `walk_ffi_safe_adts` handles `FnPtr`s, you could do

```rust
walk_ffi_safe_adts(tcx, TyKind::FnPtr(sig), &mut |did| foreign_mentioned_tys.insert(did));
```

Reply:
I don't see how I can call `walk` on a `TyKind`, I think things are fine as they are here

Reply:
Oh, oops.  You could do `tcx.mk_fn_ptr(sig)` instead of the `TYKind`.

Reply:
It does handle `FnPtrs`, see https://github.com/immunant/c2rust/pull/960#discussion_r1232403470